azkaban-common/src/main/java/azkaban/executor/container/ContainerCleanupManager.java

```diff
@@ -172,7 +172,13 @@ public void cleanUpContainersInTerminalStatuses() {
       if (!activeFlows.contains(executionId)) {
         logger.info("Cleaning up the stale pod and service for finished execution: {}",
             executionId);
-        deleteContainerQuietly(executionId);
+        try {
+          this.containerizedImpl.deletePod(executionId);
```

Comment:
Why change this? Why not keep the try catch but keep the deleteContainerQuietly API?

Reply:
I changed deleteContainerQuietly to accept a Flow Object instead of an executionId. This is so that we can interrogate the status inside the flow object down the stack. In general we should be passing as much info down the stack as we can, IMHO. Here we are relying on our own naming convention and essentially making a hard contract on something that we impose. So we have some options on the implementation:

1. We duplicate some code with a path that takes an executionId and a path that takes a Flow object. This will look a little funny, but it is workable. As part of this change we are already duplicating deletePod.
2. We change the logic here to fetch a flow for each executionId. The issue with that is then we are making a lot of calls to the DB essentially on startup of our application. This will make it so it takes longer for the WS to come up.
3. We do something like I've done here.

I think the best option might be the first one, in all honesty. The issue with what we've done here is that if an exception is always thrown for the pod then the service will never be cleaned up. In structuring the code this way I am kind of assuming that an error will not be thrown for a delete call immediately after kubernetes tells us something is there. Also, pods that are deleted from this part of the workflow won't have logs show up in the UI! 

Ultimately, as we've discussed before, I think we're making a mistake by trying to manage an orchestrator like kubernetes. I think it would make the most sense to give k8s a deployment definition and have it manage the pod/service lifecycle. There is a lot of code written by teams of people to manage edge cases we haven't even thought of. By taking on all of these responsibilities ourselves we open ourselves up to a whole  host of bugs/issues. I believe we are beginning to see some of those issues come to a head here.